src/DeviceInput/deviceInputSystem.ts

```diff
@@ -0,0 +1,223 @@
+    private _elementToAttachTo: Nullable<HTMLElement>;
+
+    /**
+     * Default Constructor
+     * @param elementToAttachTo - element to attach events to (usually canvas)
+     */
+    constructor(elementToAttachTo: Nullable<HTMLElement>) {
+        this._elementToAttachTo = elementToAttachTo;
+        this.handleKeyActions();
+        this.handleMouseActions();
+        this.handleGamepadActions();
+        this.updateDevices();
+    }
+
+    // Public functions
+    /**
+     * Checks for current device input value, given an id and input index
+     * @param deviceName Id of connected device
+     * @param inputIndex Index of device input
+     * @returns Current value of input
+     */
+    public pollInput(deviceName: string, inputIndex: number): number {
+        if (this._inputs.has(deviceName)) {
+            var device = this._inputs.get(deviceName);
+
+            if (device![inputIndex] != undefined) {
+                return device![inputIndex];
+            }
+            else {
+                throw `Unable to find input ${inputIndex} on device ${deviceName}`;
+            }
+        }
+        else {
+            throw `Unable to find device ${deviceName}`;
+        }
+    }
+
+    /**
+     * When a device is connected, perform user specified function
+     * @param callback Callback function to use when a device is connected
+     */
+    public onDeviceConnected(callback: (deviceName: string) => void): void {
+        this._onDeviceConnected = callback;
+    }
+
+    /**
+     * When a device is disconnected, perform user specified function
+     * @param callback Callback function to use when a device is disconnected
+     */
+    public onDeviceDisconnected(callback: (deviceName: string) => void): void {
+        this._onDeviceDisconnected = callback;
+    }
+
+    // Private functions
+    /**
+     * Add device and inputs to device map
+     * @param deviceName Assigned name of device (may be SN)
+     * @param numberOfInputs Number of input entries to create for given device
+     */
+    private registerDevice(deviceName: string, numberOfInputs: number) {
+        if (!this._inputs.has(deviceName)) {
+            var device: Array<number> = [];
+            for (var i = 0; i < numberOfInputs; i++) {
+                device.push(0);
+            }
+
+            this._inputs.set(deviceName, device);
+        }
+    }
+
+    /**
+     * Given a specific device name, remove that device from the device map
+     * @param deviceName Name of device to be removed
+     */
+    private deregisterDevice(deviceName: string) {
+        this._inputs.delete(deviceName);
+    }
+
+    /**
+     * Handle all actions that come from keyboard interaction
+     */
+    private handleKeyActions() {
+        window.addEventListener("keydown", (evt) => {
+            if (!this._keyboardActive) {
+                this._keyboardActive = true;
+                this.registerDevice(DeviceInputSystem.KEYBOARD_DEVICE, 222);

```

Comment:
why 222 ?

Reply:
I found that Javascript had 222 keycodes for input from the keyboard.  I didn't find a way to get that value programmatically so I used the hard value.

Reply:
what about pointer IDs? I would recommend using a negative value to be safe

Reply:
Since we're not using pointer ids, I believe that this isn't applicable.  I have pulled out the numbers into separate variables.